lib/folio/holdings.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module Folio
+  class Holdings
+    YEAR_REGEX = /(?:19|20)\d{2}/
+    # We've seen chronologies that looks like full dates, MON YYYY, WIN YYYY, and nil
+    def self.find_latest(holdings)
+      # NOTE: We saw some piece records without 'chronology'. Was this just test data?
+      pieces = holdings.filter_map do |piece|
+        piece.merge(sortable_date: extract_sortable_date(piece['chronology']) || extract_sortable_date(piece['enumeration']),
+                    sortable_tokens: extract_sortable_tokens(piece['chronology']) || extract_sortable_tokens(piece['enumeration']))
+      end
+      pieces.max_by { |piece| [piece.fetch(:sortable_date), piece.fetch(:sortable_tokens)] }
+    end
+
+    # Find any dates that are present in the fields
+    def self.extract_sortable_date(date_str)
+      return unless date_str
+
+      begin
+        return Date.parse(date_str) # handles MON YYYY
+      rescue Date::Error
+        nil
+      end
+
+      return unless (years = date_str.scan(YEAR_REGEX).presence)
+
+      max_year = years.map(&:to_i).max
+      Date.parse("#{max_year}-01-01")
+    end
+
+    # Look for other tokens, which are not dates, that can be used to sort.
+    # We need to exclude dates so this field can be used as a tie-breaker
+    def self.extract_sortable_tokens(str)
+      return unless str
+
+      tokens = str.scan(/\d+/).grep_v(YEAR_REGEX)
+      tokens.map(&:to_i).max
```

Comment:
In #905, one of the examples was "bd.54 t.2" (say, vs "bd.54 t.1'). Won't the `max` here just pull out `54` and indeterminately sort the rest?

Reply:
Ah, good catch. I've added a test for this case.